### ISSUE TITLE
CUDA compilation on ArchLinux fails

### ISSUE THREAD

**Issue by e14159** [tensorflow/tensorflow#1346]
## Overview

Build fails with 

```
/usr/include/string.h: In function 'void* __mempcpy_inline(void*, const void*, size_t)':
/usr/include/string.h:652:42: error: 'memcpy' was not declared in this scope
     return (char *) memcpy (__dest, __src, __n) + __n;
```

## Specific information:
- ArchLinux system
- Compute Capability 3.0 necessary
- CuDNN 4, CUDA 7.5
- Bazel 0.2.0

#### Followed the installation from source instructions:
- CuDNN 4, CUDA 7.5 are installed in /opt/cuda
- `git clone -b v0.7.1 --recurse-submodules https://github.com/tensorflow/tensorflow`
- `cd tensorflow`
- `./configure` Details as attached. Building with cuda support and compute capability 3.0
- `~/bin/bazel build -c opt --config=cuda //tensorflow/cc:tutorials_example_trainer`
This results in the following error:
```

/usr/lib/gcc/x86_64-unknown-linux-gnu/5.3.0/include/mwaitxintrin.h(36): error: identifier "__builtin_ia32_monitorx" is undefined

/usr/lib/gcc/x86_64-unknown-linux-gnu/5.3.0/include/mwaitxintrin.h(42): error: identifier "__builtin_ia32_mwaitx" is undefined

2 errors detected in the compilation of "/tmp/tmpxft_00006c12_00000000-7_avgpooling_op_gpu.cu.cpp1.ii".
```
This is probably due to cuda requiring gcc 4.9, while I am using 5.3. I fixed this by editing `crosstool_wrapper_driver_is_not_gcc` in `tensorflow/third_party/gpus/crosstool/clang/bin`, changing line 49 to the explicit
`GCC_HOST_COMPILER_PATH = ('/usr/bin/gcc-4.9')`
This allows compilation to continue; however, it then fails with 
```
/usr/include/string.h: In function 'void* __mempcpy_inline(void*, const void*, size_t)':
/usr/include/string.h:652:42: error: 'memcpy' was not declared in this scope
   return (char *) memcpy (__dest, __src, __n) + __n;
                                          ^
ERROR: /home/RED/src/tensorflow/tensorflow/core/BUILD:334:1: output 'tensorflow/core/_objs/gpu_kernels/tensorflow/core/kernels/aggregate_ops_gpu.cu.o' was not created.
ERROR: /home/RED/src/tensorflow/tensorflow/core/BUILD:334:1: not all outputs were created.
Target //tensorflow/cc:tutorials_example_trainer failed to build
```

Additional information:
- I had been able to compile 0.6.0 (though with CUDA 7.0 and CuDNN 2). This also necessitated switching GCC.
- CPU-Only compilation works.
- Wheel installation works, too, but is obviously unable to use my CC3.0 device.

Thanks!

Attachments:
- [tensorflow_configure.txt](https://github.com/tensorflow/tensorflow/files/153447/tensorflow_configure.txt)
- [tensorflow_output.txt](https://github.com/tensorflow/tensorflow/files/153446/tensorflow_output.txt)

**Comment by vrv** [comment]
I am not sure either.  It looks like maybe the failure is during nvcc compiling aggregate_ops_gpu.cu.cc, but I don't understand why it works for us at CUDA 7.0 and not for you.

**Comment by bchretien** [comment]
The `__builtin_ia32_monitorx`/`__builtin_ia32_mwaitx ` error is a fairly common issue. You can try disabling `mwaitx` when using nvcc (append this to your nvcc flags: `-Xcompiler;-mno-mwaitx`). Forcing `__STRICT_ANSI__` may help as well, although that's something I only had to use on my own code since it relies on Boost 1.60, and it seemed to be related to missing `__float128` support in nvcc (cf [this](https://github.com/arrayfire/arrayfire/issues/1210)).

As for your second issue, I had the same problem just a few days ago in my own code. In my case, CMake used `/usr/lib/hardening-wrapper/bin/cc` rather than `gcc` directly. That wrapper adds some extra compiler options (e.g. `_FORTIFY_SOURCE`) and can usually be used without even thinking about it, but apparently the source of the `memcpy` error was in one of these extra flags. I didn't try to pinpoint the exact source of the problem, but that's probably what you should be looking at.

**Comment by e14159** [comment]
> I am not sure either. It looks like maybe the failure is during nvcc compiling aggregate_ops_gpu.cu.cc, but I don't understand why it works for us at CUDA 7.0 and not for you.

For reference, I have also used CUDA 7.0 (and CuDNN 4). Same error.

> You can try disabling mwaitx when using nvcc (append this to your nvcc flags: -Xcompiler -mno-mwaitx). 

This was one of the other things I had experimented with before (following #1066). I ran into the second error, too.

> As for your second issue, I had the same problem just a few days ago in my own code. In my case, CMake used /usr/lib/hardening-wrapper/bin/cc rather than gcc directly. That wrapper adds some extra compiler options (e.g. _FORTIFY_SOURCE) and can usually be used without even thinking about it, but apparently the source of the memcpy error was in one of these extra flags. I didn't try to pinpoint the exact source of the problem, but that's probably what you should be looking at.

No `hardening-wrapper` installed, but looking at the flags might be a good start.

For now, I am going to roll back to 0.6.0, reproduce the exact steps I needed to get it to run (since I know it once worked) and try to apply them to 0.7.1. I'll try to keep this issue updated.

Thanks, both of you!

**Comment by bchretien** [comment]
@e14159 can you try with `-D_FORCE_INLINES`? I just had the same issue with pcl, I checked the `string.h` header, and using that preprocessor directive skips the block where the `memcpy` error appears. There's probably a cleaner workaround though...

**Comment by e14159** [comment]
It worked!
Specifically, I've inserted `cxx_flag: "-D_FORCE_INLINES"` into `tensorflow/third_party/gpus/crosstool/CROSSTOOL`:48.
I have compiled it successfully (despite some complaints due to it being both inlined and outlined) both with CUDA7.0/CuDNN2 (only with the test `bazel build -c opt --config=cuda //tensorflow/cc:tutorials_example_trainer`) and CUDA7.5/CuDNN4 (both the test and the package compilation). I have installed the latter, and ran it with an example of mine. It used the gpu, and it trained. I'm therefore, for now, going to assume that it also works correctly.

Thank you very much!

Closing this issue; I hope it'll help others running into the same issue.
